src/main/java/gregtech/loaders/recipe/RecyclingRecipes.java

```diff
@@ -1,111 +1,400 @@
-            fluidExtractorRecipeBuilder.buildAndRegister();
-        }
-
-        if (!ignoreArcSmelting) {
-            List<ItemStack> resultList = materials.stream().map(RecyclingRecipes::getArcSmeltingResult).collect(Collectors.toList());
-            resultList.removeIf(ItemStack::isEmpty);
-            if (resultList.isEmpty()) return;
-            RecipeBuilder<?> arcFurnaceRecipeBuilder = RecipeMaps.ARC_FURNACE_RECIPES.recipeBuilder()
-                    .outputs(resultList)
-                    .duration((int) Math.max(1L, firstStack.amount * 60 / M))
-                    .EUt(30 * voltageMultiplier);
-            inputSupplier.accept(arcFurnaceRecipeBuilder);
-            arcFurnaceRecipeBuilder.buildAndRegister();
+        // Call to the three recycling methods
+        registerMaceratorRecycling(input, components, voltageMultiplier);
+        registerExtractorRecycling(input, components, voltageMultiplier);
+        if (!ignoreArcSmelting) registerArcRecycling(input, components, voltageMultiplier);
+    }
+
+    private static void registerMaceratorRecycling(ItemStack input, List<MaterialStack> materials, int multiplier) {
+
+        // Finalize the output list.
+        List<ItemStack> outputs = finalizeOutputs(
+                materials,
+                RecipeMaps.MACERATOR_RECIPES.getMaxOutputs(),
+                OreDictUnifier::getDust
+        );
+
+        // Exit if no valid Materials exist for this recycling Recipe.
+        if (outputs.size() == 0) return;
+
+        // Build the final Recipe.
+        RecipeMaps.MACERATOR_RECIPES.recipeBuilder()
+                .inputs(input.copy())
+                .outputs(outputs)
+                .duration(calculateDuration(materials))
+                .EUt(4 * multiplier)
+                .buildAndRegister();
+    }
+
+    private static void registerExtractorRecycling(ItemStack input, List<MaterialStack> materials, int multiplier) {
+
+        // Find the first Material which can create a Fluid.
+        // If no Material in the list can create a Fluid, return.
+        MaterialStack fluidMs = materials.stream().filter(ms -> ms.material.hasProperty(PropertyKey.FLUID)).findFirst().orElse(null);
+        if (fluidMs == null) return;
+
+        // Find the next MaterialStack, which will be the Item output.
+        // This can sometimes be before the Fluid output in the list, so we have to
+        // assume it can be anywhere in the list.
+        MaterialStack itemMs = materials.stream().filter(ms -> !ms.material.equals(fluidMs.material)).findFirst().orElse(null);
+
+        // Calculate the duration based off of those two possible outputs.
+        // - Sum the two Material amounts together (if both exist)
+        // - Divide the sum by M
+        long duration = fluidMs.amount * fluidMs.material.getAverageMass();
+        if (itemMs != null) duration += (itemMs.amount * itemMs.material.getAverageMass());
+        duration = Math.max(1L, duration / M);
```

Comment:
Can't this just be passed into `calculateDuration` if you add some simple null checking to that method? eg: `if(material == null) continue;`

Reply:
I could, but honestly I think how it is here is as clean as it can get in the code